GoogleContainerTools/skaffold
# feat: parameterization

Fixes: #8245, #7954

 Future work:  #8306, #8406, #8410  add kpt examples

### Description: 
 - move transformers responsibility out of specific kpt render as
   - transformers and validators are KRM functions, those functions should not be tied to a specific KRM function executors(kpt) in this case. We may support running krm functions without kpt(kustomize, or internal implementation).
   - ``Kustomize`` removes comments when building manifests,  the previous approach for using transformers doesn't work for manifests with comments. 
   -  creating extra hidden directory is not necessary, also it's contaminating users' workspace
   -  ``kpt render `` command outputs Kptfile content as part of render result unless using kpt with version under 1.0.0-beta.15 which is less preferred, this actually breaks krm function specs, and will cause `kubectl apply -f -` fail if piped with the render result. 
   - support parameterizations for all renders
     - kustomize
       - as kustomize build erases comments in the source manifests and we don't want to change users' source manifests either, we mirror kustomize directory to temporary folder and do transformation(apply-setters)  in a imperative way, then run kustomize build against the transformed manifests.
     - helm 
       - using helm native support approach ``--set flag``
       - This is not working if the render config is not directly set for manifests.helm,
     - kubectl  
     - kpt 
       - no extra folder, use ``kpt render`` command and output unwrapped result into memory, then running transformers against the output.
       - This currently doesn't work if kpt version is above 1.0.0-beta.15, will fix that in another pr. 


### Manual Test 
   - add kpt setters comment in your manifests, run ``skaffold render --set key1=value1 --set key2=value2``, should be able to see values with kpt setter comments are replaced by provided values from command line.
